### PR TITLE
v3.1.0

### PR DESCRIPTION
**Pull request by TinyPandas**
## Description
Implements a new logging system, choices on slash-command options, and fixes a wide variation of bugs.

## Motivation and Context
This change is required to continue utilizing implementations of d.js to further enhance the end user experience of the bot.

Linked Issues: #59, #70, #75, #76

## How Has This Been Tested?
Testing was done on a feature-PR basis, making sure no bugs or breaking changes made their way into develop.

## Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -71,7 +71,6 @@ typings/
 
 # dotenv environment variables file
 .env
```
Comment (TinyPandas) [inline .gitignore:73]:
```suggestion

```

File: package.json [hunk 2]
```diff
@@ -1,13 +1,12 @@
 {
 	"name": "pandaprotector",
-	"version": "3.0.3",
+	"version": "3.0.4",
```
Comment (TinyPandas) [inline package.json:3]:
```suggestion
	"version": "3.1.0",
```

File: package.json [hunk 3]
```diff
@@ -1,13 +1,12 @@
...
 	"description": "RSA - Utility Bot",
 	"main": "index.js",
 	"scripts": {
 		"lint": "npx eslint . && npx prettier --check .",
 		"build": "npx tsc",
 		"start": "node dist/index.js",
-		"test": "npm run build && node dist/index.js config-test.json .env.test",
 		"check": "npm run build && npm run lint"
```
Comment (TinyPandas) [inline package.json:10]: For testing purposes. :trollface: 

```suggestion
		"test": "npm run build && npm start",
		"check": "npm run build && npm run lint"
```

File: src/commands/debug.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+const command: Command = {
+	name: "debug",
+	description: "Test logger.",
+	options: [],
+	hasPermission: (state, interaction) =>
+		(interaction.member as GuildMember).roles.cache.has(state.config.developerRoleId),
+	shouldBeEphemeral: (state, interaction) => interaction.channelID !== state.config.botChannelId,
+	handler: (state, interaction) => {
+		interaction
+			.reply("This command helps with intentional debug testing.")
```
Comment (TinyPandas) [inline src/commands/debug.ts:14]:
```suggestion
			.reply("This command helps with intentional debug testing.", { ephemeral: command.shouldBeEphemeral(state, interaction) })
```

Perhaps make this ephemeral? Maybe not, as it is dev only anyways.

File: src/config.ts [hunk 5]
```diff
@@ -8,6 +8,8 @@ export interface Config {
 	botChannelId: string;
 	staffCategoryId: string;
 	ghRepoPath: string;
+	debugMode: string;
+	debugChannelId: string;
```
Comment (TinyPandas) [inline src/config.ts:12]:
```suggestion
	logChannelId: string;
```

File: src/config.ts [hunk 6]
```diff
@@ -8,6 +8,8 @@ export interface Config {
 	botChannelId: string;
 	staffCategoryId: string;
 	ghRepoPath: string;
+	debugMode: string;
```
Comment (TinyPandas) [inline src/config.ts:11]:
```suggestion
	verbosityLevel: string;
```

File: src/config.ts [hunk 7]
```diff
@@ -24,6 +26,8 @@ export function isConfig(config: unknown): config is Config {
 		typeof record["reportChannelId"] === "string" &&
 		typeof record["botChannelId"] === "string" &&
 		typeof record["staffCategoryId"] === "string" &&
-		typeof record["ghRepoPath"] === "string"
+		typeof record["ghRepoPath"] === "string" &&
+		typeof record["debugMode"] === "string" &&
```
Comment (TinyPandas) [inline src/config.ts:30]:
```suggestion
		typeof record["verbosityLevel"] === "string" &&
```

File: src/config.ts [hunk 8]
```diff
@@ -24,6 +26,8 @@ export function isConfig(config: unknown): config is Config {
...
+		typeof record["debugChannelId"] === "string"
```
Comment (TinyPandas) [inline src/config.ts:31]:
```suggestion
		typeof record["logChannelId"] === "string"
```

File: src/index.ts [hunk 9]
```diff
@@ -1,42 +1,52 @@
...
+				})
+				.catch(err => log(err, logLevels.error));
+		})
 	);
 }
 
-function main(state: State, env: DotEnv) {
+function main(state: State, secrets: Secrets) {
 	const { config, client } = state;
+	initLogger(state, config.debugMode);
```
Comment (TinyPandas) [inline src/index.ts:44]:
```suggestion
	initLogger(state, config.verbosityLevel);
```

File: src/logger.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+	debug: { level: "DEBUG", binding: console.debug.bind(console) },
+};
+
+export function log(message: string, logLevel: LogLevel): void {
+	logLevel.binding(`[${logLevel.level}]: ${message}`);
+
+	const shouldLog = verbosityLevels.indexOf(logLevel.level.toLowerCase()) <= verbosityLevels.indexOf(verbosityLevel);
+
+	if (state && shouldLog) {
+		const logChannel = state.client.channels.resolve(state.config.debugChannelId) as TextChannel;
```
Comment (TinyPandas) [inline src/logger.ts:49]:
```suggestion
		const logChannel = state.client.channels.resolve(state.config.logChannelId) as TextChannel;
```

**Review by M39a9am3R — DISMISSED** [review]

**Review by M39a9am3R — DISMISSED** [review]

File: src/command.ts [hunk 11]
```diff
@@ -1,11 +1,10 @@
...
 export interface Command {
 	readonly name: string;
 	readonly description: string;
 	readonly options: ApplicationCommandOptionData[];
-	readonly hasPermission: (state: State, interaction: CommandInteraction) => boolean;
-	readonly shouldBeEphemeral: (state: State, interaction: CommandInteraction) => boolean;
-	readonly handler: (state: State, interaction: CommandInteraction, args: CommandInteractionOption[]) => void;
+	readonly hasPermission: (interaction: CommandInteraction) => boolean;
+	readonly shouldBeEphemeral: (interaction: CommandInteraction) => boolean;
+	readonly handler: (interaction: CommandInteraction, args: CommandInteraction["options"]) => void;
```
Comment (TinyPandas) [inline src/command.ts:9]: Replace with CommandInteractionOption[]

File: src/config.ts [hunk 12]
```diff
@@ -1,13 +1,17 @@
+import type { Snowflake } from "discord-api-types";
```
Comment (cntkillme) [inline src/config.ts:1]: Change this to importing from `discord.js` if we are using a pinned version.

**Review by cntkillme — DISMISSED** [review]

**Review by cntkillme — APPROVED** [review]

**Review by M39a9am3R — APPROVED** [review]

**Pull request merged**